README.md
# screen_adaptation

 *flutter 屏幕适配方案*
 
![手机效果](./example/dome1.jpeg)![手机效果](./example/dome2.jpeg)

![手机效果](./example/dome3.gif)
 
 
> 注意事项 最好 dart sdk>=2.6 [没有使用低版本测试如有问题请反馈]()
 
### 配置pubspec.yaml
```yaml
dependencies:
  screen_adaptation: ^{latest version}
```
### 使用的时候引入
```dart
import 'package:screen_adaptation/screen_adaptation.dart';
```
### 使用方法
```dart
class MyApp extends StatelessWidget {
  // This widget is the root of your application.  
  @override
  Widget build(BuildContext context) {
    /// [OrientationBuilder] 不需要旋转屏幕方向UI可以省略 直接返回MaterialApp
    return OrientationBuilder(
        builder: (BuildContext context, Orientation orientation) => MaterialApp(
          title: 'Flutter Demo',
          theme: ThemeData(
            primarySwatch: Colors.blue,
          ),
          home: MyHomePage(),
          builder: (BuildContext context, Widget child) {
            return ScreenAdaptationInit(
                allowFontScaling: true,
                size: Size(750, 1334),
                child: child);
          },
        ));
  }
}
```
> 如果需要旋转的屏幕的时候自动刷新UI 需要在最外层增加 OrientationBuilder 不需要请忽略
 
> 如果旋转之后页面没有刷新UI 比如使用flutter_modular的时候 可以把最外层的 OrientationBuilder去掉 在需要旋转刷新UI的外层增加 OrientationBuilder

### 参数说明
- ScreenAdaptationInit

|属性|类型|默认值|描述|
|:---|:---|:---|:---|
|size|Size|Size(750, 1334)|设计稿中设备的尺寸|
|landscapeSize|Size|null|横屏UI设计图尺寸 可不传 默认为 竖屏[size]尺寸反转 [landscapeSize.width]为[size.height] [landscapeSize.height]为[size.width]|
|allowFontScaling|bool|false|字体是否根据系统缩放比例变化|

- ScreenAdaptationUtil

*这里使用screenAdaptationUtil 代替 ScreenAdaptationUtil.getInstance()*

|属性|类型|使用方式|描述|
|:---|:---|:---|:---|
|width|double|screenAdaptationUtil.width|设计宽度
|height|double|screenAdaptationUtil.height|设计高度
|screenWidth|double|screenAdaptationUtil.screenWidth|屏幕宽度
|screenHeight|double|screenAdaptationUtil.screenHeight|屏幕高度
|pixelRatio|double|screenAdaptationUtil.pixelRatio|屏幕密度
|textScaleFactor|double|screenAdaptationUtil.textScaleFactor|系统字体缩放比例
|statusBarHeight|double|screenAdaptationUtil.statusBarHeight|状态栏高度
|bottomBarHeight|double|screenAdaptationUtil.bottomBarHeight|底部安全距离
|allowFontScaling|bool|screenAdaptationUtil.allowFontScaling|是否允许字体根据系统缩放 
|setWidth|Function|screenAdaptationUtil.setWidth(num)或者[num].w|根据宽度进行适配 
|setHeight|Function|screenAdaptationUtil.setHeight(num)或者[num].h|根据高度进行适配 
|setSp|Function|screenAdaptationUtil.setSp(fontSize,allowFontScaling:true/false)或者[fontSize].s [fontSize].st [fontSize].sf|文字适配  allowFontScaling 是否允许系统缩放
|setAutomatic|Function|screenAdaptationUtil.setAutomatic(num)或者[num].r|根据宽高比例较小的适配 

### 例子
```dart
class DomePage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: SingleChildScrollView(
        padding: EdgeInsets.only(
            top: ScreenAdaptationUtil.getInstance().statusBarHeight + 30.w,
            left: 30.w,
            right: 30.w,
            bottom: 30.w),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            Wrap(
              spacing: 30.w,
              runSpacing: 30.w,
              children: [
                Container(
                  width: ScreenAdaptationUtil.getInstance().setWidth(100),
                  height: ScreenAdaptationUtil.getInstance().setWidth(100),
                  color: Colors.blueGrey,
                  alignment: Alignment.center,
                ),
                Container(
                  width: 100.h,
                  height: 100.w,
                  color: Colors.blueGrey,
                  alignment: Alignment.center,
                ),
                Container(
                  width: 100.w,
                  height: 100.h,
                  color: Colors.blueGrey,
                  alignment: Alignment.center,
                ),
                Container(
                  width: 100.h,
                  height: 100.h,
                  color: Colors.blueGrey,
                  alignment: Alignment.center,
                ),
                Container(
                  width: 100.r,
                  height: 100.r,
                  color: Colors.blueGrey,
                  alignment: Alignment.center,
                ),
              ],
            ),
            Container(
              child: Text('设计宽度:${ScreenAdaptationUtil.getInstance().width}'),
            ),
            Container(
              child: Text('设计高度:${ScreenAdaptationUtil.getInstance().height}'),
            ),
            Container(
              child: Text(
                  '屏幕宽度:${ScreenAdaptationUtil.getInstance().screenWidth}'),
            ),
            Container(
              child: Text(
                  '屏幕高度:${ScreenAdaptationUtil.getInstance().screenHeight}'),
            ),
            Container(
              child:
                  Text('屏幕密度:${ScreenAdaptationUtil.getInstance().pixelRatio}'),
            ),
            Container(
              child: Text(
                  '文字缩放比例:${ScreenAdaptationUtil.getInstance().textScaleFactor}'),
            ),
            Container(
              child: Text(
                  '状态栏高度:${ScreenAdaptationUtil.getInstance().statusBarHeight}'),
            ),
            Container(
              child: Text(
                  '底部安全距离:${ScreenAdaptationUtil.getInstance().bottomBarHeight}'),
            ),
            Container(
              child: Text(
                  '是否允许字体根据系统缩放变化:${ScreenAdaptationUtil.getInstance().allowFontScaling}'),
            ),
            Container(
              margin: EdgeInsets.only(top: 30.w),
              child: Text(
                '不会根据系统缩放变化的文字',
                style: TextStyle(
                    fontSize: ScreenAdaptationUtil.getInstance()
                        .setSp(60, allowFontScaling: false)),
              ),
            ),
            Container(
              margin: EdgeInsets.only(top: 30.w),
              child: Text(
                '不会根据系统缩放变化的文字',
                style: TextStyle(fontSize: 60.sf),
              ),
            ),
            Container(
              margin: EdgeInsets.symmetric(vertical: 30.w),
              child: Text(
                '因为ScreenAdaptationInit allowFontScaling默认为false 我也不会变化',
                style: TextStyle(fontSize: 60.s),
              ),
            ),
            Container(
              child: Text(
                '会根据系统缩放变化的文字 缩放了 ${ScreenAdaptationUtil.getInstance().textScaleFactor}',
                style: TextStyle(
                    color: Colors.blueGrey,
                    fontSize: ScreenAdaptationUtil.getInstance()
                        .setSp(60, allowFontScaling: true)),
              ),
            ),
            Container(
              child: Text(
                '会根据系统缩放变化的文字 缩放了 ${ScreenAdaptationUtil.getInstance().textScaleFactor}',
                style: TextStyle(color: Colors.blueGrey, fontSize: 60.st),
              ),
            ),
          ],
        ),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}
```


CHANGELOG.md
## [0.0.1] - TODO: Add release date.

* TODO: Describe initial release.

## [0.2.0-nullsafety.0] - TODO: Add release date.

* TODO: Null safety

## [0.2.1] - TODO: Add release date.

* TODO: Null safety




lib/screen_adaptation_init.dart
import 'package:flutter/widgets.dart';
import 'package:screen_adaptation/screen_adaptation_util.dart';

/// [child]
/// [size] UI设计图尺寸 默认750 1334
/// [landscapeSize] 横屏UI设计图尺寸 可不传 默认为 竖屏[size]尺寸反转 [landscapeSize.width]为[size.height] [landscapeSize.height]为[size.width]
/// [allowFontScaling] 字体是否根据系统缩放比例变化 默认为false
class ScreenAdaptationInit extends StatelessWidget {
  final Widget child;
  final Size size;
  final Size? landscapeSize;
  final bool allowFontScaling;

  ScreenAdaptationInit(
      {required this.child,
      this.size = const Size(750, 1337),
      this.allowFontScaling = false,
      this.landscapeSize});

  @override
  Widget build(BuildContext context) {
    return OrientationBuilder(
        builder: (BuildContext context, Orientation orientation) {
      final mediaQuery = MediaQuery.of(context);
      ScreenAdaptationUtil.instance = ScreenAdaptationUtil(
          width: size.width,
          height: size.height,
          landscapeWidth: landscapeSize?.width,
          landscapeHeight: landscapeSize?.height,
          allowFontScaling: allowFontScaling,
          orientation: orientation,
          screenWidth: mediaQuery.size.width,
          screenHeight: mediaQuery.size.height,
          statusBarHeight: mediaQuery.padding.top,
          bottomBarHeight: mediaQuery.padding.bottom,
          textScaleFactor: mediaQuery.textScaleFactor,
          pixelRatio: mediaQuery.devicePixelRatio);
      return child;
    });
  }
}


lib/screen_adaptation_util.dart
import 'dart:math';

import 'package:flutter/widgets.dart';
import 'package:screen_adaptation/screen_adaptation_provider.dart';

/// [width] 设计尺寸宽度
/// [width] 设计尺寸高度
/// [screenWidth] //屏幕宽度
/// [screenHeight] //屏幕高度
/// [orientation] //屏幕方向
/// [pixelRatio]  //像素比
/// [statusBarHeight] //距离顶部安全距离
/// [bottomBarHeight] //距离底部安全距离
/// [textScaleFactor] //文字缩放比例
/// [allowFontScaling] //是否允许文字缩放
class ScreenAdaptationUtil implements ScreenAdaptationProvider {
  late double width;
  late double height;
  double? screenWidth;
  double? screenHeight;
  Orientation? orientation;
  double? pixelRatio;
  double? statusBarHeight;
  double? bottomBarHeight;
  double? textScaleFactor;
  bool allowFontScaling;

  static ScreenAdaptationUtil? _adaptationUtil;

  static set instance(ScreenAdaptationUtil adaptationUtil) {
    _adaptationUtil = adaptationUtil;
  }

  ScreenAdaptationUtil({
    double width = 750,
    double height = 1334,
    this.screenWidth,
    this.screenHeight,
    this.orientation,
    this.pixelRatio,
    this.statusBarHeight,
    this.bottomBarHeight,
    this.textScaleFactor,
    this.allowFontScaling = false,
    double? landscapeWidth,
    double? landscapeHeight,
  }) {
    if (orientation == Orientation.portrait) {
      this.width = width;
      this.height = height;
    } else {
      this.width = landscapeWidth ?? height;
      this.height = landscapeHeight ?? width;
    }
  }

  static ScreenAdaptationUtil? getInstance() {
    if (_adaptationUtil == null) {
      assert(
        _adaptationUtil != null,
        '\n请先调用ScreenAdaptationInit()进行实例化',
      );
      return ScreenAdaptationUtil();
    }
    return _adaptationUtil;
  }

  @override
  double setAutomatic(num wh) {
    // TODO: implement setAutomatic
    return wh * min(screenWidth! / this.width, screenHeight! / this.height);
  }

  @override
  double setHeight(num height) {
    return height * (screenHeight! / this.height);
  }

  @override
  double setSp(num fontSize, {bool? allowFontScaling}) {
    if (allowFontScaling == null) {
      if (this.allowFontScaling == false)
        return fontSize *
            min(screenWidth! / this.width, screenHeight! / this.height);
      return fontSize *
          min(screenWidth! / this.width, screenHeight! / this.height) *
          textScaleFactor!;
    }
    if (allowFontScaling == false)
      return fontSize *
          min(screenWidth! / this.width, screenHeight! / this.height);
    return fontSize *
        min(screenWidth! / this.width, screenHeight! / this.height) *
        textScaleFactor!;
  }

  @override
  double setWidth(num width) {
    // TODO: implement setWidth
    return width * (screenWidth! / this.width);
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:screen_adaptation/screen_adaptation.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return OrientationBuilder(
        builder: (BuildContext context, Orientation orientation) => MaterialApp(
              title: 'Flutter Demo',
              theme: ThemeData(
                primarySwatch: Colors.blue,
              ),
              home: MyHomePage(),
              builder: (BuildContext context, Widget child) {
                return ScreenAdaptationInit(
                    size: Size(750, 1334), child: child);
              },
            ));
  }
}

class MyHomePage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    final bool _isPortrait =
        ScreenAdaptationUtil.getInstance().orientation == Orientation.portrait;
    return Scaffold(
      appBar: AppBar(
        title: Text('screen_adaptation'),
        actions: [
          TextButton(
              onPressed: () => Navigator.push(
                  context,
                  MaterialPageRoute(
                      builder: (BuildContext context) => DomePage())),
              child: Text(
                'dome',
                style: TextStyle(fontSize: 48.s, color: Colors.white),
              ))
        ],
      ),
      body: SingleChildScrollView(
        child: Column(
          children: [
            Padding(
              padding: EdgeInsets.all(30.w),
              child: Text(
                '旋转屏幕后UI会自动进行适配',
                style: TextStyle(fontSize: 48.s),
              ),
            ),
            Container(
              width: _isPortrait ? 750.w : 1334.w,
              height: 100.w,
              color: Colors.blueGrey,
              alignment: Alignment.center,
              child: Text(
                '宽度100%',
                style: TextStyle(color: Colors.white, fontSize: 35.s),
              ),
            ),
            Container(
              margin: EdgeInsets.only(top: 30.w),
              width: _isPortrait ? 345.w : 667.w,
              height: _isPortrait ? 667.h : 345.w,
              color: Colors.blueGrey,
              alignment: Alignment.center,
              child: Text(
                '宽度50% 高度50%',
                style: TextStyle(color: Colors.white, fontSize: 35.s),
              ),
            ),
          ],
        ),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}

class DomePage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: SingleChildScrollView(
        padding: EdgeInsets.only(
            top: ScreenAdaptationUtil.getInstance().statusBarHeight + 30.w,
            left: 30.w,
            right: 30.w,
            bottom: 30.w),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            Wrap(
              spacing: 30.w,
              runSpacing: 30.w,
              children: [
                Container(
                  width: ScreenAdaptationUtil.getInstance().setWidth(100),
                  height: ScreenAdaptationUtil.getInstance().setWidth(100),
                  color: Colors.blueGrey,
                  alignment: Alignment.center,
                ),
                Container(
                  width: 100.h,
                  height: 100.w,
                  color: Colors.blueGrey,
                  alignment: Alignment.center,
                ),
                Container(
                  width: 100.w,
                  height: 100.h,
                  color: Colors.blueGrey,
                  alignment: Alignment.center,
                ),
                Container(
                  width: 100.h,
                  height: 100.h,
                  color: Colors.blueGrey,
                  alignment: Alignment.center,
                ),
                Container(
                  width: 100.r,
                  height: 100.r,
                  color: Colors.blueGrey,
                  alignment: Alignment.center,
                ),
              ],
            ),
            Container(
              child: Text('设计宽度:${ScreenAdaptationUtil.getInstance().width}'),
            ),
            Container(
              child: Text('设计高度:${ScreenAdaptationUtil.getInstance().height}'),
            ),
            Container(
              child: Text(
                  '屏幕宽度:${ScreenAdaptationUtil.getInstance().screenWidth}'),
            ),
            Container(
              child: Text(
                  '屏幕高度:${ScreenAdaptationUtil.getInstance().screenHeight}'),
            ),
            Container(
              child:
                  Text('屏幕密度:${ScreenAdaptationUtil.getInstance().pixelRatio}'),
            ),
            Container(
              child: Text(
                  '文字缩放比例:${ScreenAdaptationUtil.getInstance().textScaleFactor}'),
            ),
            Container(
              child: Text(
                  '状态栏高度:${ScreenAdaptationUtil.getInstance().statusBarHeight}'),
            ),
            Container(
              child: Text(
                  '底部安全距离:${ScreenAdaptationUtil.getInstance().bottomBarHeight}'),
            ),
            Container(
              child: Text(
                  '是否允许字体根据系统缩放变化:${ScreenAdaptationUtil.getInstance().allowFontScaling}'),
            ),
            Container(
              margin: EdgeInsets.only(top: 30.w),
              child: Text(
                '不会根据系统缩放变化的文字',
                style: TextStyle(
                    fontSize: ScreenAdaptationUtil.getInstance()
                        .setSp(60, allowFontScaling: false)),
              ),
            ),
            Container(
              margin: EdgeInsets.only(top: 30.w),
              child: Text(
                '不会根据系统缩放变化的文字',
                style: TextStyle(fontSize: 60.sf),
              ),
            ),
            Container(
              margin: EdgeInsets.symmetric(vertical: 30.w),
              child: Text(
                '因为ScreenAdaptationInit allowFontScaling默认为false 我也不会变化',
                style: TextStyle(fontSize: 60.s),
              ),
            ),
            Container(
              child: Text(
                '会根据系统缩放变化的文字 缩放了 ${ScreenAdaptationUtil.getInstance().textScaleFactor}',
                style: TextStyle(
                    color: Colors.blueGrey,
                    fontSize: ScreenAdaptationUtil.getInstance()
                        .setSp(60, allowFontScaling: true)),
              ),
            ),
            Container(
              child: Text(
                '会根据系统缩放变化的文字 缩放了 ${ScreenAdaptationUtil.getInstance().textScaleFactor}',
                style: TextStyle(color: Colors.blueGrey, fontSize: 60.st),
              ),
            ),
          ],
        ),
      ), // This trailing comma makes auto-formatting nicer for build methods.
      bottomNavigationBar: TextButton(
        onPressed: () => Navigator.maybePop(context),
        child: Container(
          height: 100.w,
          alignment: Alignment.center,
          color: Colors.blue,
          child: Text(
            '返回',
            style: TextStyle(color: Colors.white, fontSize: 40.s),
          ),
        ),
      ),
    );
  }
}
